scanner/nodeFramework.go

```diff
@@ -0,0 +1,175 @@
+	// first ensure that there is a package.json
+	if !checksPass(sourceDir, fileExists("package.json")) {
+		return nil, nil
+	}
+
+	// ensure package.json has a start script
+	data, err := os.ReadFile("package.json")
+
+	if err != nil {
+		return nil, nil
+	} else {
+		err = json.Unmarshal(data, &packageJson)
+		if err != nil {
+			return nil, nil
+		}
+
+		scripts, ok := packageJson["scripts"].(map[string]interface{})
+
+		if !ok || scripts["start"] == nil {
+			return nil, nil
+		}
+	}
+
+	// ensure node version is at least 16
+	out, err := exec.Command("node", "-v").Output()
+	if err != nil {
+		return nil, nil
+	} else {
+		nodeVersion := strings.TrimSpace(string(out))
+		if nodeVersion[:1] == "v" {
+			nodeVersion = nodeVersion[1:]
+		}
+		if nodeVersion < "16" {
+			return nil, nil
+		}
+	}
+
+	srcInfo := &SourceInfo{
+		Family:     "NodeJS",
+		SkipDeploy: true,
+		Callback:   NodeFrameworkCallback,
+	}
+
+	return srcInfo, nil
+}
+
+func NodeFrameworkCallback(srcInfo *SourceInfo, options map[string]bool) error {
+	// generate Dockerfile if it doesn't already exist
+	_, err := os.Stat("Dockerfile")
+	if errors.Is(err, fs.ErrNotExist) {
+		var args []string
+
+		_, err = os.Stat("node_modules")
+		if errors.Is(err, fs.ErrNotExist) {
+			// no existing node_modules directory: run package directly
+			args = []string{"npx", "--yes", "@flydotio/dockerfile@latest"}
+		} else {
+			// build command to install package using preferred package manager
+			args = []string{"npm", "install", "@flydotio/dockerfile", "--save-dev"}
+
+			_, err = os.Stat("yarn.lock")
+			if !errors.Is(err, fs.ErrNotExist) {
+				args = []string{"yarn", "add", "@flydotio/dockerfile", "--dev"}
+			}
+
+			_, err = os.Stat("pnpm-lock.yaml")
+			if !errors.Is(err, fs.ErrNotExist) {
+				args = []string{"pnpm", "add", "-D", "@flydotio/dockerfile"}
+			}
+		}
+
+		// check first to see if the package is already installed
+		installed := false
+
+		deps, ok := packageJson["dependencies"].(map[string]interface{})
+		if ok && deps["@flydotio/dockerfile"] != nil {
+			installed = true
+		}
+
+		deps, ok = packageJson["devDependencies"].(map[string]interface{})
+		if ok && deps["@flydotio/dockerfile"] != nil {
+			installed = true
+		}
+
+		// install/run command
+		if !installed || args[0] == "npx" {
+			cmd := exec.Command(args[0], args[1:]...)
+			cmd.Stdin = nil
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+
+			if err := cmd.Run(); err != nil {
+				return errors.Wrap(err, "Failed to install @flydotio/dockerfile")
+			}
+		}
+
+		// run the package if we haven't already
+		if args[0] != "npx" {
+			cmd := exec.Command("npx", "dockerfile")
+			cmd.Stdin = nil
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+
+			if err := cmd.Run(); err != nil {
+				return errors.Wrap(err, "Failed to generate Dockerfile")
+			}
+		}
+	}
+
+	// read dockerfile
+	dockerfile, err := os.ReadFile("Dockerfile")
```

Comment:
The [dockerfile-node](https://github.com/fly-apps/dockerfile-node) generator has plenty of tests, including actually using Docker to build each framework.  What other tests do you have in mind?

Reply:
Some tests around the regexps. Since we don't have a full-fledged Dockerfile parser here, this code may need some iterations to handle all Dockerfile in the wild.

Reply:
Also I just like shorter functions :)

Reply:
I'm not totally following, but it seems to me that this can be a follow-on to this change.

The regex logic for parsing out the port is the same that I added to the Dockrfile scanner.  If you write tests for that scanner, perhaps I can backport them to this scanner?

Reply:
Yeah. It could be a follow-up change.